java/src/jmri/jmrit/vsdecoder/VSDecoderManager.java

```diff
@@ -540,32 +532,26 @@ protected void shutdownDecoders() {
         Iterator<String> it = vk.iterator();
         while (it.hasNext()) {
             VSDecoder v = decodertable.get(it.next());
+
+            if (v.getEngineSound().isEngineStarted()) {
+                log.warn("Shutting down a running Engine ...");
+                v.getEngineSound().setEngineStarted(false);
+                snooze(1500);
```

Comment:
Can you explain what you are doing here? A delay of 1500 milliseconds seems to be a long time. If there are several locos which are active, it seems to me that the total delay might be very long.

Is this running on the GUI thread (same as the layout thread) or on a separate thread?

Reply:
The challenge and weak point is: closing the VSDManager window with (maximum 4) running engines (throttle > 0). In all other cases, even on JMRI Quit, 300ms delay is enough.
I think the VSDManager window is a GUI thread. There are also an engine loop player thread (queued sounds) and several audio threads (command, loop sound, move sound, fadeOut sound). These threads can continue running after a sound is stopped. Will do more tests. Would 300ms instead of 1500ms a compromise?

The error I get (sometimes) at a small delay time is an audio object state error. VSD and JMRI remain operable.

Reply:
Where is `shutDownDecoders()` called from?  If it's on the main or GUI thread, we need to find another way to do this. Perhaps spinning this off to another thread?

There might be a more fundamental issue, though.  (Or perhaps I've misunderstood the structure or naming). Closing a window should _not_ be destroying a manager.  They're separate parts of the MVC structure.  To put it another way, if you close the window, why should the sounds stop?

Reply:
> Where is shutDownDecoders() called from? If it's on the main or GUI thread, we need to find another way to do this. Perhaps spinning this off to another thread?
>
> There might be a more fundamental issue, though. (Or perhaps I've misunderstood the structure or naming). Closing a window should not be destroying a manager. They're separate parts of the MVC structure. To put it another way, if you close the window, why should the sounds stop?

Interesting comment. Thank you!
I think the design is almost unchanged since I continued the VSD author's work. Let's see if I understood. There is a VSDecoderManager thread (main thread). There is a VSDManagerFrame (GUI thread). When this window is closed, `shutDownDecoders()` is called. This stops all the sounds. Next step is to set the VSDManagerFrame to null. I'm afraid this is not conform to the MVC approach.